src/viewer/NativePdfViewer.jsx

```diff
@@ -0,0 +1,146 @@
+/* global cozy cordova */
+import React, { Component } from 'react'
+import {
+  temporarySave,
+  getNativeFile,
+  openFileWithCordova
+} from 'drive/mobile/lib/filesystem'
+import NoViewer from './NoViewer'
+import Spinner from 'cozy-ui/react/Spinner'
+import classNames from 'classnames'
+import Button from 'cozy-ui/react/Button'
+import { translate } from 'cozy-ui/react/I18n'
+
+import styles from './styles'
+
+const VIEWER_OPTIONS = {
+  email: {
+    enabled: true
+  },
+  print: {
+    enabled: true
+  },
+  openWith: {
+    enabled: true
+  },
+  bookmarks: {
+    enabled: true
+  },
+  search: {
+    enabled: false
+  },
+  autoClose: {
+    onPause: false
+  }
+}
+
+class NativePdfViewer extends Component {
+  state = {
+    opening: true,
+    error: null
+  }
+
+  nativeFileUrl = null
+
+  onOpen = () => {
+    this.setState({ opening: false })
+  }
+
+  onClose = () => {
+    this.props.onClose()
+  }
+
+  onError = e => {
+    console.warn(e)
+    this.setState({ error: e })
+  }
+
+  componentWillMount = async () => {
+    try {
+      if (this.props.file.isAvailableOffline) {
+        const file = await getNativeFile(this.props.file)
+        this.nativeFileUrl = file.nativeURL
+      } else {
+        const response = await cozy.client.files.downloadById(
+          this.props.file.id
+        )
+        const blob = await response.blob()
+        const file = await temporarySave(blob, this.props.file.name)
+        this.nativeFileUrl = file.nativeURL
+      }
+
+      this.openPdfInExternalViewer()
+    } catch (e) {
+      this.onError(e)
+    }
+  }
+
+  openPdfInExternalViewer = () => {
+    cordova.plugins.SitewaertsDocumentViewer.viewDocument(
```

Comment:
Tu dois tester si on a bien le plugin, et qu'on est bien dans cordova, etc.
Tu trouveras des exemples de code dans le répertoire https://github.com/cozy/cozy-drive/tree/master/src/drive/mobile/lib
L'idée de ce répertoire est de mettre tout le code **cordova** dans un même endroit pour n'exposer que des API qui servent de contrat et ainsi on a un couplage moins fort entre notre app et les fonctionnalités natives.

Reply:
Je vais en effet le déplacer, pour le couplage et par homogénéité. Ceci dit, la verification cordova est faite en amont dans le `Viewer` — le NativePdfViewer ne peut que s'afficher dans cordova. Et si le plugin n'est pas présent, le try catch autour du code appelant s'occupe de ce qu'il faut.

Reply:
En fait je suis même pas sur que ce soit une bonne idée. Le viewer est pensé comme un duck "pure", il ne repose pas sur drive — c'est pour ca qu'il redéfini par exemple [`isCordova`](https://github.com/cozy/cozy-drive/blob/master/src/viewer/Viewer.jsx#L22). Si je fais un `import` depuis le code spécifique de drive, je met à mal cette séparation.

Un avis @goldoraf ?

Reply:
Effectivement, j'ai vu ça dans https://github.com/cozy/cozy-drive/blob/master/src/viewer/Viewer.jsx
Alors qu'il existe ça https://github.com/cozy/cozy-drive/blob/master/src/drive/mobile/lib/device.js

Les deux implémentations ne sont pas les mêmes : 

### dans `Viewer.jsx`

```js
const isAndroid = () =>
  window.navigator.userAgent &&
  window.navigator.userAgent.indexOf('Android') >= 0
const isIOS = () =>
  window.navigator.userAgent &&
  /iPad|iPhone|iPod/.test(window.navigator.userAgent)
const isMobile = () => isAndroid() || isIOS()
const isCordova = () => window.cordova !== undefined
```

### dans `device.js`

```js
export const isCordova = () => window.cordova !== undefined
export const getPlatformId = () =>
  isCordova() ? window.cordova.platformId : undefined
const isPlatform = platform => getPlatformId() === platform
export const isIos = () => isPlatform('ios')
export const isAndroid = () => isPlatform('android')
```

Ce choix de réimplémenter pour ne pas dépendre d'un autre fichier est un choix. Ce n'est pas cela qu'on a appelé `ducks` : [The Ducks File Structure for Redux](https://medium.com/@scbarrus/the-ducks-file-structure-for-redux-d63c41b7035c)
Dans cet article "fondateur", on ne parle pas de `React`, on ne parle pas de composant, on ne parle que de Redux. Redux n'a rien à voir avec React ou les composants.

Ce choix de réimplémenter pour ne pas dépendre d'un autre fichier est un choix. C'est le choix d'avoir un lot de composants standalone. Mais si on ne veut pas dépendre de nos petits helpers cordova, dans ce cas, il faut que ce lot ne dépende pas d'autres objets du projet.
Or je vois des `import { ImageLoader } from 'components/Image'` ou autre `import { isCordova } from 'drive/mobile/lib/device'` et `import { openOfflineFile, openOnlineFile } from 'drive/mobile/lib/filesystem'`.

Donc le choix n'est pas net, il faut choisir. Jusqu'à nouvel ordre, le choix est de conserver l'architecture de fichiers actuelle et de mettre tout ce qui dépend de cordova dans une arborescence à part. Tout en proposant une API, un contrat, qui permet de découpler le choix d'implémentation (après tout `isAndroid` peut être implementer de multiple manière) du besoin fonctionnel (savoir qu'on est sur Android).

Avoir à plusieurs endroits les mêmes implémentations ne m'apparait pas un choix pérenne et facilement maintenable (DRY).

Mais si on décide de changer, car ce n'est pas parce que c'est comme ça que c'est la bonne solution, il n'y a aucun problème à tout revoir et à changer d'avis sur des choix passés, alors il faut le faire bien. En modifiant l'ancien code avec les nouvelles décisions.

Ce sont des choix d'organisation de fichiers, d'architecture et de conception. Et tous les choix se justifient. Mais l'important c'est de partager ses choix et que ces choix soient partagés/acceptés par tous afin d'être homogène et de ne pas découvrir une nouvelle façon de faire à chaque arborescence technico-fonctionnelle.

Reply:
Pour la postérité : comme convenu de vive-vois, on s'est mis d'accord sur la direction à prendre et je l'ai consignée dans un fichier README ici.